Rename DebugAccessDecisionManager to TraceableAccessDecisionManager

Jean85 opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #21085
| License       | MIT

[EDIT] No longer WIP, test passing. Also, test added to preserve BC with the SecurityBundle.

Jean85 committed:
Rename DebugAccessDecisionManager to TraceableAccessDecisionManager (#21085)

wouterj commented:
Tests seem to pass on Travis. So maybe a local problem? I think you're correct about not having to care about BC here as it's `@internal`.
 
status: reviewed

xabbuh reviewed (changes requested):
We need to keep the old class name (possibly through a class alias). Otherwise, older versions of the SecurityBundle would break when being used with 3.3/3.4 versions of the Security Core component. We should probably add a test case (the fact that the build suite passed indicates that we are not testing anything/the right thing here).

Jean85 commented:
Well, it was not a local error but a fail of the Travis build on my fork, here: https://travis-ci.org/Jean85/symfony/builds/187412622

If we need a BC class, I will add it. In fact, mixing versions of the bundles may break stuff like in my build, you're right!

Jean85 committed:
Add deprecated DebugAccessDecisionManager class alias

xabbuh commented on src/Symfony/Component/Security/Core/Authorization/DebugAccessDecisionManager.php:
```diff
@@ -1,98 +1,18 @@
+ *  with older SecurityBundle versions. 
  */
-class DebugAccessDecisionManager implements AccessDecisionManagerInterface
+class DebugAccessDecisionManager extends TraceableAccessDecisionManager
```
What about using the same approach that @nicolas-grekas used in #20967 instead?

Jean85 replied:
Nice! Doing it RN!

Jean85 replied:
Test failing locally... Maybe I'm missing something, but how is it supposed to work like that? It's defining the alias before declaring the class, it says that the Traceable class is missing!

xabbuh replied:
Can you push your latest commit so we can take a look at it?

Jean85 replied:
Pushed!

Jean85 committed:
Attempt at following example from #20967 for class rename and deprecation

xabbuh commented on src/Symfony/Component/Security/Core/Authorization/TraceableAccessDecisionManager.php:
```diff
@@ -0,0 +1,100 @@
+
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+
+class_alias(TraceableAccessDecisionManager::class, DebugAccessDecisionManager::class);
```
moving this line after the class definition should make tests green

Jean85 replied:
I've chosen a different approach: I've removed the `class_exist` (since it seems useless to me) and I've moved it into the deprecated file, which seems the right place to me.

Jean85 committed:
Fix class_alias moving it to the deprecated class file

Jean85 committed:
Add assertion to test instanceof of deprecated class

Jean85 committed:
Add test for reverse BC preservation

Jean85 commented:
Latests commits are coming from the discussion with @nicolas-grekas in #21096.

xabbuh commented:
@Jean85 In your test, the `DebugAccessDecisionManager` class was already loaded in line 47.

Jean85 commented:
You're right; I've tested locally and it works even if I launch the single tests method, so it's just a test issue, to avoid future regressions.

Maybe it's better if I move the tests to the top of the class, and I use a simple `class_exist`, to avoid this issue?

xabbuh commented:
@Jean85 Not sure I completely understood what you meant with your last question. I suggest to simply follow the approach from #20967.

Jean85 committed:
Fix test for BC, revert to suggested approach

Jean85 commented:
I've reverted to the suggested approach, an rewrote the test to avoid triggering the autoload of the deprecated class.

Sorry if I was so stubborn @xabbuh , now I've rewritten the test to show how my approach was flawed. PHPUnit's `assertInstanceOf` was doing a `class_exist` behind the scene, triggering the autoload and hiding my mistake.

xabbuh commented on src/Symfony/Component/Security/Core/Tests/Authorization/TraceableAccessDecisionManagerTest.php:
```diff
@@ -40,4 +41,16 @@ public function provideObjectsAndLogs()
     }
+
+    /**
+     * @group this
```
should be removed

xabbuh commented:
@Jean85 Don't worry about that. Renaming is actually not so easy and it's always good to think about the approaches we use.

Jean85 committed:
Remove unused group annotation on test

Jean85 commented:
Thanks! I've removed the annotation that I left behind by mistake!

nicolas-grekas commented on src/Symfony/Component/Security/Core/Authorization/DebugAccessDecisionManager.php:
```diff
@@ -1,98 +1,15 @@
- * about the security voters and the decisions made by them.
- *
- * @author Javier Eguiluz <javier.eguiluz@gmail.com>
+ * @author Alessandro Lai <alessandro.lai85@gmail.com>
```
you should remove this line, there is no authorship to claim here: git will record your contribution

nicolas-grekas commented on src/Symfony/Component/Security/Core/Authorization/DebugAccessDecisionManager.php:
```diff
@@ -1,98 +1,15 @@
+ * @author Alessandro Lai <alessandro.lai85@gmail.com>
  *
  * @internal
+ * @deprecated The DebugAccessDecisionManager class is deprecated since version 3.3 and will be removed in 4.0. Use the Symfony\Component\Security\Core\Authorization\TraceableAccessDecisionManager class instead.
```
the namespace could be removed

nicolas-grekas commented on src/Symfony/Component/Security/Core/Tests/Authorization/TraceableAccessDecisionManagerTest.php:
```diff
@@ -40,4 +41,13 @@ public function provideObjectsAndLogs()
+    {
+        $adm = new TraceableAccessDecisionManager(new AccessDecisionManager());
+
+        if (! $adm instanceof DebugAccessDecisionManager) {
```
extra space to be removed after `!`

nicolas-grekas commented on src/Symfony/Component/Security/Core/Tests/Authorization/TraceableAccessDecisionManagerTest.php:
```diff
@@ -40,4 +41,13 @@ public function provideObjectsAndLogs()
+        $adm = new TraceableAccessDecisionManager(new AccessDecisionManager());
+
+        if (! $adm instanceof DebugAccessDecisionManager) {
+            $this->fail('Doesn\'t work, BC with SecurityBundle broken, see PR #21088');
```
we don't reference github, `git blame` is used instead to look for historical things. I propose:
`$this->fail('For BC, TraceableAccessDecisionManager must be an instance of DebugAccessDecisionManager');`

nicolas-grekas reviewed (commented):
:+1: with minor changes

Jean85 committed:
Minor fixes to comply to review

Jean85 commented:
Fixes committed! The style checker is having some false positive I think...

xabbuh commented on src/Symfony/Component/Security/Core/Authorization/DebugAccessDecisionManager.php:
```diff
@@ -1,98 +1,13 @@
- *
- * For the full copyright and license information, please view the LICENSE
- * file that was distributed with this source code.
- */
```
This should IMO not be removed.

Jean85 committed:
Restoring copyright notice

Jean85 commented:
@xabbuh done!

xabbuh commented:
👍 

Status: Reviewed

nicolas-grekas commented:
Thank you @Jean85.
